File: typescript/tests/conformance.test.ts
/**
 * Cross-SDK conformance tests for the TypeScript SDK.
 *
 * These tests run against a mock server started by the conformance orchestrator.
 * They are automatically skipped when ENSOUL_CONFORMANCE_URL is not set,
 * so regular `vitest run` is unaffected.
 */
import { describe, it, expect, afterAll } from "vitest";
import { Ensoul } from "../src/client.js";
import {
  NotFoundError,
  AuthenticationError,
  AuthorizationError,
  RateLimitError,
  ValidationError,
  ServerError,
} from "../src/errors.js";
import { parseChatEvent, type ChatStreamEvent } from "../src/streaming.js";
import type { HTTPClient } from "../src/http.js";

const CONFORMANCE_URL = process.env.ENSOUL_CONFORMANCE_URL;
const describeConformance = CONFORMANCE_URL ? describe : describe.skip;

describeConformance("Conformance Tests", () => {
  const client = new Ensoul({
    apiKey: "sk_test_123",
    baseUrl: CONFORMANCE_URL!,
    maxRetries: 0,
    customHeaders: { "X-SDK-Language": "typescript" },
  });

  const noAuthClient = new Ensoul({
    apiKey: "",
    baseUrl: CONFORMANCE_URL!,
    maxRetries: 0,
  });

  afterAll(() => {
    client.close();
    noAuthClient.close();
  });

  // ---------------------------------------------------------------------------
  // Personas
  // ---------------------------------------------------------------------------

  describe("Personas", () => {
    it("test_persona_create", async () => {
      const persona = await client.personas.create({
        name: "Test Persona",
        domain: "test_domain",
        personalityData: { trait_a: 75, trait_b: 50 },
      });
      expect(persona.id).toBe("p_test_001");
      expect(persona.name).toBe("Test Persona");
      expect(persona.domain).toBe("test_domain");
    });

    it("test_persona_get", async () => {
      const persona = await client.personas.get("p_test_001");
      expect(persona.id).toBe("p_test_001");
      expect(persona.name).toBe("Test Persona");
      expect(persona.domain).toBe("test_domain");
    });

    it("test_persona_update", async () => {
      const persona = await client.personas.update("p_test_001", {
        name: "Updated Persona",
        personalityData: { trait_a: 80, trait_b: 60 },
      });
      expect(persona.id).toBe("p_test_001");
      expect(persona.name).toBe("Updated Persona");
    });

    it("test_persona_delete", async () => {
      await expect(
        client.personas.delete("p_test_001"),
      ).resolves.toBeUndefined();
    });

    it("test_persona_list_pagination", async () => {
      const page = await client.personas.list({ page: 1, perPage: 10 });
      expect(page.items.length).toBeGreaterThanOrEqual(1);
      expect(page.total).toBe(25);
      expect(page.page).toBe(1);
      expect(page.perPage).toBe(10);
      expect(page.pages).toBe(3);
    });

    it("test_persona_not_found", async () => {
      await expect(
        client.personas.get("nonexistent_persona_id"),
      ).rejects.toThrow(NotFoundError);

      try {
        await client.personas.get("nonexistent_persona_id");
      } catch (err) {
        expect(err).toBeInstanceOf(NotFoundError);
        expect((err as NotFoundError).statusCode).toBe(404);
      }
    });
  });

  // ---------------------------------------------------------------------------
  // Chat
  // ---------------------------------------------------------------------------

  describe("Chat", () => {
    it("test_chat_send", async () => {
      const response = await client.chat.send("p_test_001", "Hello, how are you?");
      expect(response.response).toBeTruthy();
      expect(response.conversation_id).toBeTruthy();
      expect(response.token_usage.total_tokens).toBeGreaterThan(0);
    });

    it("test_chat_stream_sse", async () => {
      const stream = await client.chat.stream("p_test_001", "Tell me about yourself.");
      const events: ChatStreamEvent[] = [];

      for await (const sseEvent of stream) {
        events.push(parseChatEvent(sseEvent));
      }

      expect(events).toHaveLength(5);

      // Check chunk ordering
      for (let i = 0; i < events.length; i++) {
        expect(events[i]!.chunkIndex).toBe(i);
      }

      // Final event
      const lastEvent = events[events.length - 1]!;
      expect(lastEvent.isFinal).toBe(true);
      expect(lastEvent.tokenUsage).toBeDefined();
      expect(lastEvent.tokenUsage!.totalTokens).toBeGreaterThan(0);

      // Non-final events
      for (const event of events.slice(0, -1)) {
        expect(event.isFinal).toBe(false);
      }
    });

    it("test_chat_get_conversations", async () => {
      const page = await client.chat.getConversations("p_test_001");
      expect(page.items.length).toBeGreaterThanOrEqual(1);
      expect(page.total).toBe(2);
    });
  });

  // ---------------------------------------------------------------------------
  // Domains
  // ---------------------------------------------------------------------------

  describe("Domains", () => {
    it("test_domain_list", async () => {
      const page = await client.domains.list();
      expect(page.items.length).toBeGreaterThanOrEqual(1);
    });

    it("test_domain_get", async () => {
      const domain = await client.domains.get("d_test_001");
      expect(domain.id).toBe("d_test_001");
      expect(domain.name).toBe("Test Domain");
    });
  });

  // ---------------------------------------------------------------------------
  // Simulations
  // ---------------------------------------------------------------------------

  describe("Simulations", () => {
    it("test_simulation_create", async () => {
      const sim = await client.simulations.create({
        name: "Test Simulation",
        domainId: "d_test_001",
      });
      expect(sim.id).toBe("sim_test_001");
      expect(sim.status).toBe("created");
    });

    it("test_simulation_start", async () => {
      const result = await client.simulations.start("sim_test_001", { ticks: 50 });
      expect(result.status).toBe("running");
      expect(result.ticks_requested).toBe(50);
    });
  });

  // ---------------------------------------------------------------------------
  // Memory
  // ---------------------------------------------------------------------------

  describe("Memory", () => {
    it("test_memory_create", async () => {
      const mem = await client.memory.create("p_test_001", {
        content: "Remembered something important",
        source: "user",
      });
      expect(mem.id).toBe("mem_test_001");
    });

    it("test_memory_delete", async () => {
      await expect(
        client.memory.delete("p_test_001", "mem_test_001"),
      ).resolves.toBeUndefined();
    });
  });

  // ---------------------------------------------------------------------------
  // Sessions
  // ---------------------------------------------------------------------------

  describe("Sessions", () => {
    it("test_session_create", async () => {
      const session = await client.sessions.create({ tier: 0 });
      expect(session.id).toBe("sess_test_001");
      expect(session.tier).toBe(0);
      expect(session.parent_session_id).toBeNull();
    });
  });

  // ---------------------------------------------------------------------------
  // Aggregate
  // ---------------------------------------------------------------------------

  describe("Aggregate", () => {
    it("test_aggregate_count", async () => {
      const result = await client.aggregate.count({ domain: "demo" });
      expect(result.sample_size).toBe(500);
      expect(result.confidence).toBe(0.95);
    });
  });

  // ---------------------------------------------------------------------------
  // Health
  // ---------------------------------------------------------------------------

  describe("Health", () => {
    it("test_health_check", async () => {
      const result = await client.health.check();
      expect(result.status).toBe("ok");
      expect(result.version).toBeTruthy();
    });
  });

  // ---------------------------------------------------------------------------
  // Info
  // ---------------------------------------------------------------------------

  describe("Info", () => {
    it("test_info_config", async () => {
      const result = await client.info.config();
      expect(result.api_version).toBe("1.0.0");
      expect(result.max_batch_size).toBe(100);
    });
  });

  // ---------------------------------------------------------------------------
  // Auth Resources
  // ---------------------------------------------------------------------------

  describe("Auth Resources", () => {
    it("test_auth_token_exchange", async () => {
      const token = await client.auth.token("testuser", "testpass");
      expect(token.access_token).toBeTruthy();
      expect(token.token_type).toBe("bearer");
      expect(token.expires_in).toBe(3600);
    });

    it("test_auth_me", async () => {
      const user = await client.auth.me();
      expect(user.consumer_id).toBe("user_test_001");
      expect(user.username).toBe("testuser");
    });
  });

  // ---------------------------------------------------------------------------
  // Frameworks
  // ---------------------------------------------------------------------------

  describe("Frameworks", () => {
    it("test_framework_update", async () => {
      const fw = await client.frameworks.update("fw_test_001", {
        name: "Big Five Updated",
      });
      expect(fw.id).toBe("fw_test_001");
      expect(fw.name).toBe("Big Five Updated");
    });
  });

  // ---------------------------------------------------------------------------
  // Errors
  // ---------------------------------------------------------------------------

  describe("Errors", () => {
    it("test_error_rate_limit", async () => {
      const rateLimitClient = new Ensoul({
        apiKey: "sk_test_123",
        baseUrl: CONFORMANCE_URL!,
        maxRetries: 0,
        customHeaders: { "X-Trigger-RateLimit": "true" },
      });

      try {
        await expect(rateLimitClient.personas.list()).rejects.toThrow(RateLimitError);

        try {
          await rateLimitClient.personas.list();
        } catch (err) {
          expect(err).toBeInstanceOf(RateLimitError);
          expect((err as RateLimitError).retryAfter).toBe(30);
        }
      } finally {
        rateLimitClient.close();
      }
    });

    it("test_error_validation", async () => {
      // Access the internal HTTP client to POST an empty body to /v1/personas
      const httpClient = (client as unknown as { _client: HTTPClient })._client;

      await expect(
        httpClient.post("/v1/personas", {}),
      ).rejects.toThrow(ValidationError);

      try {
        await httpClient.post("/v1/personas", {});
      } catch (err) {
        expect(err).toBeInstanceOf(ValidationError);
        expect((err as ValidationError).details.length).toBeGreaterThanOrEqual(1);
      }
    });

    it("test_error_authentication", async () => {
      await expect(noAuthClient.personas.list()).rejects.toThrow(AuthenticationError);

      try {
        await noAuthClient.personas.list();
      } catch (err) {
        expect(err).toBeInstanceOf(AuthenticationError);
        expect((err as AuthenticationError).statusCode).toBe(401);
      }
    });

    it("test_error_server", async () => {
      const serverErrorClient = new Ensoul({
        apiKey: "sk_test_123",
        baseUrl: CONFORMANCE_URL!,
        maxRetries: 0,
        customHeaders: { "X-Trigger-ServerError": "true" },
      });

      try {
        await expect(serverErrorClient.personas.list()).rejects.toThrow(ServerError);

        try {
          await serverErrorClient.personas.list();
        } catch (err) {
          expect(err).toBeInstanceOf(ServerError);
          expect((err as ServerError).statusCode).toBe(500);
        }
      } finally {
        serverErrorClient.close();
      }
    });

    it("test_error_authorization_forbidden", async () => {
      const forbiddenClient = new Ensoul({
        apiKey: "sk_test_123",
        baseUrl: CONFORMANCE_URL!,
        maxRetries: 0,
        customHeaders: { "X-Trigger-Forbidden": "true" },
      });

      try {
        await expect(forbiddenClient.personas.list()).rejects.toThrow(AuthorizationError);

        try {
          await forbiddenClient.personas.list();
        } catch (err) {
          expect(err).toBeInstanceOf(AuthorizationError);
          expect((err as AuthorizationError).statusCode).toBe(403);
        }
      } finally {
        forbiddenClient.close();
      }
    });

    it("test_error_retry_503", async () => {
      const retryClient = new Ensoul({
        apiKey: "sk_test_123",
        baseUrl: CONFORMANCE_URL!,
        maxRetries: 2,
        customHeaders: {
          "X-Trigger-503-Once": "true",
          "X-SDK-Language": "typescript-retry",
        },
      });

      try {
        const page = await retryClient.personas.list();
        expect(page.items.length).toBeGreaterThanOrEqual(1);
      } finally {
        retryClient.close();
      }
    });
  });

  // ---------------------------------------------------------------------------
  // Auth
  // ---------------------------------------------------------------------------

  describe("Auth", () => {
    it("test_auth_api_key_header", async () => {
      // If we can list personas successfully, the auth header was correct
      const page = await client.personas.list();
      expect(page.items.length).toBeGreaterThanOrEqual(1);
    });

    it("test_auth_no_credentials", async () => {
      await expect(noAuthClient.personas.list()).rejects.toThrow(AuthenticationError);
    });

    it("test_auth_bearer_token", async () => {
      const bearerClient = new Ensoul({
        bearerToken: "test_token_123",
        baseUrl: CONFORMANCE_URL!,
        maxRetries: 0,
      });

      try {
        const page = await bearerClient.personas.list();
        expect(page.items.length).toBeGreaterThanOrEqual(1);
      } finally {
        bearerClient.close();
      }
    });
  });

  // ---------------------------------------------------------------------------
  // Chat sessions (persisted history)
  // ---------------------------------------------------------------------------

  describe("Chat Sessions", () => {
    it("test_create_session", async () => {
      const session = await client.chat.createSession({
        teamId: "team_test_001",
        userId: "user_test_001",
        domainId: "d_test_001",
        personaId: "persona_test_001",
        title: "Test Chat Session",
      });
      expect(session.id).toBe("csess_test_001");
      expect(session.is_archived).toBe(false);
    });

    it("test_list_sessions", async () => {
      const result = await client.chat.listSessions({ userId: "user_test_001" });
      const sessions = result.sessions as unknown[];
      const pagination = result.pagination as Record<string, unknown>;
      expect(sessions.length).toBeGreaterThanOrEqual(1);
      expect(pagination.total).toBe(1);
    });

    it("test_session_stats", async () => {
      const result = await client.chat.sessionStats({
        teamId: "team_test_001",
        startDate: "2025-01-01",
        endDate: "2025-01-31",
      });
      expect(result.total).toBe(7);
    });

    it("test_get_session", async () => {
      const session = await client.chat.getSession("csess_test_001");
      expect(session.id).toBe("csess_test_001");
      expect((session.messages as unknown[]).length).toBeGreaterThanOrEqual(1);
    });

    it("test_update_session", async () => {
      const session = await client.chat.updateSession("csess_test_001", {
        title: "Renamed",
      });
      expect(session.id).toBe("csess_test_001");
    });

    it("test_archive_session", async () => {
      const session = await client.chat.archiveSession("csess_test_001");
      expect(session.id).toBe("csess_test_001");
    });

    it("test_delete_session", async () => {
      await expect(
        client.chat.deleteSession("csess_test_001"),
      ).resolves.toBeUndefined();
    });

    it("test_add_message", async () => {
      const message = await client.chat.addMessage("csess_test_001", {
        role: "assistant",
        content: "Hi",
      });
      expect(message.id).toBe("msg_test_002");
      expect(message.role).toBe("assistant");
    });

    it("test_get_messages", async () => {
      const messages = await client.chat.getMessages("csess_test_001");
      expect(messages).toHaveLength(2);
      expect(messages[0]!.role).toBe("user");
    });
  });

  // ---------------------------------------------------------------------------
  // Simulation participants and event ticks
  // ---------------------------------------------------------------------------

  describe("Simulation Participants", () => {
    it("test_list_participants", async () => {
      const result = await client.simulations.listParticipants("sim_test_001");
      expect(result.total).toBe(2);
      expect((result.items as unknown[]).length).toBe(2);
    });

    it("test_add_participants", async () => {
      const sim = await client.simulations.addParticipants("sim_test_001", [
        "persona_test_001",
      ]);
      expect(sim.id).toBe("sim_test_001");
    });

    it("test_event_ticks", async () => {
      const result = await client.simulations.getEventTicks("sim_test_001");
      expect((result.ticks as unknown[]).length).toBe(3);
    });
  });

  // ---------------------------------------------------------------------------
  // Audit and verification
  // ---------------------------------------------------------------------------

  describe("Audit", () => {
    it("test_get_event", async () => {
      const event = await client.audit.getEvent("evt_test_001");
      expect(event.event_id).toBe("evt_test_001");
      expect(event.event_hash).toBeTruthy();
    });

    it("test_get_commitment", async () => {
      const commitment = await client.audit.getCommitment("cmt_test_001");
      expect(commitment.commitment_id).toBe("cmt_test_001");
      expect(commitment.event_count).toBe(42);
    });

    it("test_get_proof", async () => {
      const proof = await client.audit.getProof("evt_test_001");
      expect(proof.verified).toBe(true);
      expect((proof.proof_path as unknown[]).length).toBe(2);
    });

    it("test_verify", async () => {
      const result = await client.audit.verify("evt_test_001");
      expect(result.verified).toBe(true);
    });

    it("test_signing_key", async () => {
      const pem = await client.audit.getSigningKey();
      expect(pem).toContain("BEGIN PUBLIC KEY");
    });
  });

  // ---------------------------------------------------------------------------
  // Pagination
  // ---------------------------------------------------------------------------

  describe("Pagination", () => {
    it("test_pagination_auto_fetch", async () => {
      const page = await client.frameworks.list({ perPage: 2 });
      const allItems: Record<string, unknown>[] = [];

      for await (const item of page.autoPagingIter()) {
        allItems.push(item);
      }

      expect(allItems).toHaveLength(3);
    });
  });

  // ---------------------------------------------------------------------------
  // Client Configuration
  // ---------------------------------------------------------------------------

  describe("Client Configuration", () => {
    it("test_client_custom_base_url", async () => {
      // Verify the client respects custom baseUrl by connecting to mock server
      const customClient = new Ensoul({
        apiKey: "sk_test_123",
        baseUrl: CONFORMANCE_URL!,
        maxRetries: 0,
      });

      try {
        const page = await customClient.personas.list();
        expect(page.items.length).toBeGreaterThanOrEqual(1);
      } finally {
        customClient.close();
      }
    });
  });
});

File: kotlin/src/test/kotlin/ai/ensoul/sdk/ConformanceTest.kt
/**
 * Cross-SDK conformance tests for the Kotlin SDK.
 *
 * These tests run against a mock server started by the conformance orchestrator.
 * They are automatically skipped when ENSOUL_CONFORMANCE_URL is not set,
 * so regular `./gradlew test` runs are unaffected.
 */
package ai.ensoul.sdk

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.ints.shouldBeGreaterThan
import io.kotest.matchers.ints.shouldBeGreaterThanOrEqual
import io.kotest.matchers.shouldBe
import io.kotest.matchers.shouldNotBe
import io.kotest.matchers.string.shouldContain
import io.kotest.matchers.string.shouldNotBeBlank
import io.ktor.http.*
import kotlinx.coroutines.flow.toList
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.double
import kotlinx.serialization.json.int
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

private val CONFORMANCE_URL: String? = System.getenv("ENSOUL_CONFORMANCE_URL")

class ConformanceTest : FunSpec({

    // Skip the entire spec when the conformance URL is not configured.
    if (!CONFORMANCE_URL.isNullOrBlank()) {

    // Shared client with valid API key auth, retries disabled.
    val client by lazy {
        EnsoulClient(
            apiKey = "sk_test_123",
            baseUrl = CONFORMANCE_URL!!,
            maxRetries = 0,
            customHeaders = mapOf("X-SDK-Language" to "kotlin"),
        )
    }

    // Client with no credentials for auth-failure tests.
    val noAuthClient by lazy {
        EnsoulClient(
            apiKey = "",
            baseUrl = CONFORMANCE_URL!!,
            maxRetries = 0,
        )
    }

    afterSpec {
        client.close()
        noAuthClient.close()
    }

    // -----------------------------------------------------------------------
    // Personas
    // -----------------------------------------------------------------------

    test("persona create") {
        val persona = client.personas.create(
            name = "Test Persona",
            domain = "test_domain",
            personalityData = mapOf("trait_a" to 75, "trait_b" to 50),
        )
        persona.id shouldBe "p_test_001"
        persona.name shouldBe "Test Persona"
        persona.domain shouldBe "test_domain"
    }

    test("persona get") {
        val persona = client.personas.get("p_test_001")
        persona.id shouldBe "p_test_001"
        persona.name shouldBe "Test Persona"
        persona.domain shouldBe "test_domain"
    }

    test("persona update") {
        val persona = client.personas.update(
            "p_test_001",
            fields = mapOf("name" to "Updated Persona"),
        )
        persona.name shouldBe "Updated Persona"
    }

    test("persona delete") {
        // DELETE returns void — no exception means success.
        client.personas.delete("p_test_001")
    }

    test("persona list pagination") {
        val page = client.personas.list(page = 1, perPage = 10)
        page.items.size shouldBeGreaterThan 0
        page.total shouldBe 25
        page.page shouldBe 1
        page.perPage shouldBe 10
        page.pages shouldBe 3
    }

    test("persona not found") {
        val error = shouldThrow<NotFoundError> {
            client.personas.get("nonexistent_persona_id")
        }
        error.statusCode shouldBe 404
    }

    // -----------------------------------------------------------------------
    // Chat
    // -----------------------------------------------------------------------

    test("chat send") {
        val response = client.chat.send("p_test_001", "Hello, how are you?")
        response.response.shouldNotBeBlank()
        response.conversationId.shouldNotBeBlank()
        response.tokenUsage.totalTokens shouldBeGreaterThan 0
    }

    test("chat stream SSE") {
        val stream = client.chat.stream("p_test_001", "Tell me about yourself.")
        val sseEvents = stream.events().toList()
        val chatEvents = sseEvents
            .filter { it.event == "chunk" }
            .map { parseChatEvent(it) }

        chatEvents.size shouldBe 5

        // Check chunk ordering
        chatEvents.forEachIndexed { i, event ->
            event.chunkIndex shouldBe i
            event.conversationId shouldBe "conv_stream_001"
        }

        // Final event
        val last = chatEvents.last()
        last.isFinal shouldBe true
        last.tokenUsage shouldNotBe null
        last.tokenUsage!!["total_tokens"]!! shouldBeGreaterThan 0

        // Non-final events
        chatEvents.dropLast(1).forEach { event ->
            event.isFinal shouldBe false
        }
    }

    test("chat get conversations") {
        val page = client.chat.getConversations("p_test_001")
        page.items.size shouldBeGreaterThanOrEqual 1
        page.total shouldBe 2
    }

    // -----------------------------------------------------------------------
    // Domains
    // -----------------------------------------------------------------------

    test("domain list") {
        val page = client.domains.list()
        page.items.size shouldBeGreaterThan 0
    }

    test("domain get") {
        val domain = client.domains.get("d_test_001")
        domain["id"]!!.jsonPrimitive.contentOrNull shouldBe "d_test_001"
        domain["name"]!!.jsonPrimitive.contentOrNull shouldBe "Test Domain"
    }

    // -----------------------------------------------------------------------
    // Simulations
    // -----------------------------------------------------------------------

    test("simulation create") {
        val sim = client.simulations.create(
            name = "Test Simulation",
            domainId = "d_test_001",
        )
        sim.id shouldBe "sim_test_001"
        sim.status.name.lowercase() shouldBe "created"
    }

    test("simulation start") {
        val result = client.simulations.start("sim_test_001", ticks = 50)
        result["status"]!!.jsonPrimitive.contentOrNull shouldBe "running"
        result["ticks_requested"]!!.jsonPrimitive.int shouldBe 50
    }

    // -----------------------------------------------------------------------
    // Memory
    // -----------------------------------------------------------------------

    test("memory create") {
        val mem = client.memory.create(
            personaId = "p_test_001",
            content = "Test memory content",
        )
        mem["id"]!!.jsonPrimitive.contentOrNull shouldBe "mem_test_001"
    }

    test("memory delete") {
        // DELETE returns void — no exception means success.
        client.memory.delete("p_test_001", "mem_test_001")
    }

    // -----------------------------------------------------------------------
    // Sessions
    // -----------------------------------------------------------------------

    test("session create") {
        val session = client.sessions.create()
        session["id"]!!.jsonPrimitive.contentOrNull shouldBe "sess_test_001"
        session["tier"]!!.jsonPrimitive.int shouldBe 0
        session["parent_session_id"]!! shouldBe kotlinx.serialization.json.JsonNull
    }

    // -----------------------------------------------------------------------
    // Aggregate
    // -----------------------------------------------------------------------

    test("aggregate count") {
        val result = client.aggregate.count(domain = "test_domain")
        result["sample_size"]!!.jsonPrimitive.int shouldBe 500
        result["confidence"]!!.jsonPrimitive.double shouldBe 0.95
    }

    // -----------------------------------------------------------------------
    // Health
    // -----------------------------------------------------------------------

    test("health check") {
        val result = client.health.check()
        result["status"]!!.jsonPrimitive.contentOrNull shouldBe "ok"
    }

    // -----------------------------------------------------------------------
    // Info
    // -----------------------------------------------------------------------

    test("info config") {
        val result = client.info.config()
        result["api_version"]!!.jsonPrimitive.contentOrNull shouldBe "1.0.0"
        result["max_batch_size"]!!.jsonPrimitive.int shouldBe 100
    }

    // -----------------------------------------------------------------------
    // Auth Resources
    // -----------------------------------------------------------------------

    test("auth token exchange") {
        val token = client.auth.token(username = "testuser", password = "testpass")
        token.accessToken.shouldNotBeBlank()
        token.tokenType shouldBe "bearer"
    }

    test("auth me") {
        val user = client.auth.me()
        user.consumerId shouldBe "user_test_001"
    }

    // -----------------------------------------------------------------------
    // Frameworks
    // -----------------------------------------------------------------------

    test("framework update") {
        val fw = client.frameworks.update(
            "fw_test_001",
            body = mapOf("name" to "Big Five Updated"),
        )
        fw["id"]!!.jsonPrimitive.contentOrNull shouldBe "fw_test_001"
        fw["name"]!!.jsonPrimitive.contentOrNull shouldBe "Big Five Updated"
    }

    // -----------------------------------------------------------------------
    // Errors
    // -----------------------------------------------------------------------

    test("error rate limit") {
        val error = shouldThrow<RateLimitError> {
            client.httpClient.request(
                method = HttpMethod.Get,
                path = "/v1/personas",
                headers = mapOf("X-Trigger-RateLimit" to "true"),
            )
        }
        error.retryAfter shouldBe 30
    }

    test("error validation") {
        val error = shouldThrow<ValidationError> {
            client.httpClient.post("/v1/personas", json = emptyMap())
        }
        error.statusCode shouldBe 422
        error.details.size shouldBeGreaterThan 0
    }

    test("error authentication") {
        val error = shouldThrow<AuthenticationError> {
            noAuthClient.personas.list()
        }
        error.statusCode shouldBe 401
    }

    test("error server") {
        val error = shouldThrow<ServerError> {
            client.httpClient.request(
                method = HttpMethod.Get,
                path = "/v1/personas",
                headers = mapOf("X-Trigger-ServerError" to "true"),
            )
        }
        error.statusCode shouldBe 500
    }

    test("error authorization forbidden") {
        val error = shouldThrow<AuthorizationError> {
            client.httpClient.request(
                method = HttpMethod.Get,
                path = "/v1/personas",
                headers = mapOf("X-Trigger-Forbidden" to "true"),
            )
        }
        error.statusCode shouldBe 403
    }

    test("error retry 503") {
        val retryClient = EnsoulClient(
            apiKey = "sk_test_123",
            baseUrl = CONFORMANCE_URL!!,
            maxRetries = 2,
            customHeaders = mapOf("X-SDK-Language" to "kotlin"),
        )
        try {
            // First request gets 503, retry succeeds with persona list.
            val page = retryClient.httpClient.request(
                method = HttpMethod.Get,
                path = "/v1/personas",
                headers = mapOf("X-Trigger-503-Once" to "true"),
            )
            // If we get here without exception, the retry succeeded.
            page.status.value shouldBe 200
        } finally {
            retryClient.close()
        }
    }

    // -----------------------------------------------------------------------
    // Auth
    // -----------------------------------------------------------------------

    test("auth api key header") {
        // If we can list personas successfully, the X-Api-Key header was accepted.
        val page = client.personas.list()
        page.items.size shouldBeGreaterThan 0
    }

    test("auth no credentials") {
        shouldThrow<AuthenticationError> {
            noAuthClient.personas.list()
        }
    }

    test("auth bearer token") {
        val bearerClient = EnsoulClient(
            bearerToken = "test_token_123",
            baseUrl = CONFORMANCE_URL!!,
            maxRetries = 0,
        )
        try {
            val page = bearerClient.personas.list()
            page.items.size shouldBeGreaterThan 0
        } finally {
            bearerClient.close()
        }
    }

    // -----------------------------------------------------------------------
    // Client Configuration
    // -----------------------------------------------------------------------

    test("client custom base url") {
        // Verify the client respects custom baseUrl by connecting to the mock server.
        val customClient = EnsoulClient(
            apiKey = "sk_test_123",
            baseUrl = CONFORMANCE_URL!!,
            maxRetries = 0,
        )
        try {
            val page = customClient.personas.list()
            page.items.size shouldBeGreaterThan 0
        } finally {
            customClient.close()
        }
    }

    // -----------------------------------------------------------------------
    // Pagination
    // -----------------------------------------------------------------------

    test("pagination auto fetch") {
        val page = client.frameworks.list(perPage = 2)
        val allItems = page.autoPagingFlow().toList()
        allItems.size shouldBe 3
    }

    // -----------------------------------------------------------------------
    // Chat sessions (persisted history)
    // -----------------------------------------------------------------------

    test("create session") {
        val session = client.chat.createSession(
            teamId = "team_test_001",
            userId = "user_test_001",
            domainId = "d_test_001",
            personaId = "persona_test_001",
            title = "Test Chat Session",
        )
        session["id"]!!.jsonPrimitive.contentOrNull shouldBe "csess_test_001"
        session["is_archived"]!!.jsonPrimitive.boolean shouldBe false
    }

    test("list sessions") {
        val result = client.chat.listSessions(userId = "user_test_001")
        result["sessions"]!!.jsonArray.size shouldBeGreaterThanOrEqual 1
        result["pagination"]!!.jsonObject["total"]!!.jsonPrimitive.int shouldBe 1
    }

    test("session stats") {
        val result = client.chat.sessionStats(
            teamId = "team_test_001",
            startDate = "2025-01-01",
            endDate = "2025-01-31",
        )
        result["total"]!!.jsonPrimitive.int shouldBe 7
    }

    test("get session") {
        val session = client.chat.getSession("csess_test_001")
        session["id"]!!.jsonPrimitive.contentOrNull shouldBe "csess_test_001"
        session["messages"]!!.jsonArray.size shouldBeGreaterThanOrEqual 1
    }

    test("update session") {
        val session = client.chat.updateSession("csess_test_001", title = "Renamed")
        session["id"]!!.jsonPrimitive.contentOrNull shouldBe "csess_test_001"
    }

    test("archive session") {
        val session = client.chat.archiveSession("csess_test_001")
        session["id"]!!.jsonPrimitive.contentOrNull shouldBe "csess_test_001"
    }

    test("delete session") {
        // 204 No Content — no exception means success.
        client.chat.deleteSession("csess_test_001")
    }

    test("add message") {
        val message = client.chat.addMessage(
            "csess_test_001",
            role = "assistant",
            content = "Hi",
        )
        message["id"]!!.jsonPrimitive.contentOrNull shouldBe "msg_test_002"
        message["role"]!!.jsonPrimitive.contentOrNull shouldBe "assistant"
    }

    test("get messages") {
        val messages = client.chat.getMessages("csess_test_001")
        messages.size shouldBe 2
        messages[0]["role"]!!.jsonPrimitive.contentOrNull shouldBe "user"
    }

    // -----------------------------------------------------------------------
    // Simulation participants and event ticks
    // -----------------------------------------------------------------------

    test("list participants") {
        val result = client.simulations.listParticipants("sim_test_001")
        result["total"]!!.jsonPrimitive.int shouldBe 2
        result["items"]!!.jsonArray.size shouldBe 2
    }

    test("add participants") {
        val sim = client.simulations.addParticipants("sim_test_001", listOf("persona_test_001"))
        sim["id"]!!.jsonPrimitive.contentOrNull shouldBe "sim_test_001"
    }

    test("event ticks") {
        val result = client.simulations.getEventTicks("sim_test_001")
        result["ticks"]!!.jsonArray.size shouldBe 3
    }

    // -----------------------------------------------------------------------
    // Audit and verification
    // -----------------------------------------------------------------------

    test("audit get event") {
        val event = client.audit.getEvent("evt_test_001")
        event["event_id"]!!.jsonPrimitive.contentOrNull shouldBe "evt_test_001"
        event["event_hash"]!!.jsonPrimitive.content.shouldNotBeBlank()
    }

    test("audit get commitment") {
        val commitment = client.audit.getCommitment("cmt_test_001")
        commitment["commitment_id"]!!.jsonPrimitive.contentOrNull shouldBe "cmt_test_001"
        commitment["event_count"]!!.jsonPrimitive.int shouldBe 42
    }

    test("audit get proof") {
        val proof = client.audit.getProof("evt_test_001")
        proof["verified"]!!.jsonPrimitive.boolean shouldBe true
        proof["proof_path"]!!.jsonArray.size shouldBe 2
    }

    test("audit verify") {
        val result = client.audit.verify("evt_test_001")
        result["verified"]!!.jsonPrimitive.boolean shouldBe true
    }

    test("audit signing key") {
        val pem = client.audit.getSigningKey()
        pem shouldContain "BEGIN PUBLIC KEY"
    }

    } // if CONFORMANCE_URL
})
